### ISSUE TITLE
constant attr_key?

### ISSUE THREAD

**Issue by joyousrabbit** [google/seq2seq#156]
Hello, this is a very inspired and helpful project. I have a question about calculating the attention score. 
In your attention.py, you have:
scores = self.score_fn(att_keys, att_query)
where:
att_keys = tf.contrib.layers.fully_connected(
        inputs=keys,
        num_outputs=self.params["num_units"],
        activation_fn=None,
        scope="att_keys")

but the keys from encoder are never change after encoding, and the w_att_keys will also be fixed after training. So for every step in decoding, the att_keys is constant, right? Is it meaningful to put that constant in calculating the scores? Maybe only att_query is enought?

Hope reply.

**Comment by dennybritz** [comment]
Yes, you're right, `att_keys` never changes after encoding. I previously thought that Tensorflow is smart enough to optimize it inside the loop and doesn't re-compute it every time, but after talking to some people it turns out it does actually re-compute it. So, ideally, the `att_keys` computation should be moved outside of the loop.

**Comment by joyousrabbit** [comment]
@dennybritz Thanks for fast reply. Can we only use att_key to initialize the very first attention scores, removing it from the following steps (as many other projects do)? From intuition, the network can learn a constant by itself to compensate the constant att_key. What do you think?

Thanks.
